HopsanGenerator/src/HopsanGeneratorLib.cpp

```diff
@@ -336,3 +337,248 @@ bool callExeExportGenerator(const char* outputPath, void* pHopsanSystem, const c
 }
 
+
+//! @brief Adds a component to an existing library
+//! @param[in] libraryXmlPath Absolute path to library XML file
+//! @param[in] librarySourcePath Path to library CPP file relative to path for library XML file
+//! @param[in] typeName Typename for new component
+//! @param[in] displayName Display name for new component
+bool callAddComponentToLibrary(const char* libraryXmlPath, const char* typeName, const char* displayName, const char* cqsType,
+                               const char* const constantNames[], const int numConstantNames,
+                               const char* const constantDescriptions[], const int numConstantDescriptions,
+                               const char* const constantUnits[], const int numConstantUnits,
+                               const char* const constantInits[], const int numConstantInits,
+                               const char* const inputNames[], const int numInputNames,
+                               const char* const inputDescriptions[], const int numInputDescriptions,
+                               const char* const inputUnits[], const int numInputUnits,
+                               const char* const inputInits[], const int numInputInits,
+                               const char* const outputNames[], const int numOutputNames,
+                               const char* const outputDescriptions[], const int numOutputDescriptions,
+                               const char* const outputUnits[], const int numOutputUnits,
+                               const char* const outputInits[], const int numOutputInits,
+                               const char* const portNames[], const int numPortNames,
+                               const char* const portDescriptions[], const int numPortDescriptions,
+                               const char* const portTypes[], const int numPortTypes,
+                               const int portsRequired[], const int numPortsRequired,
+                               messagehandler_t messageHandler, void* pMessageObject)
+{
+    QFileInfo xmlPath(libraryXmlPath);
+    QString cafPath = xmlPath.absoluteDir().absoluteFilePath(QString(typeName)+".xml");
+    QString hppPath = xmlPath.absoluteDir().absoluteFilePath(QString(typeName)+".hpp");
+
+    QString dummy;
+    auto pGenerator = std::unique_ptr<HopsanGeneratorBase>(new HopsanGeneratorBase(dummy, dummy));
+    pGenerator->setMessageHandler(messageHandler, pMessageObject);
+
+    //Generate CAF file for new component
+    ComponentAppearanceSpecification cafSpec(typeName);
+    cafSpec.mDisplayName = displayName;
+    cafSpec.mSourceCode = QFileInfo(hppPath).fileName();
+    cafSpec.mRecompilable = true;
+    if(!pGenerator->generateCafFile(cafPath, cafSpec)) {
+        pGenerator->printErrorMessage("Failed to generate component appearance file.");
+        return false;
+    };
+
+    //Generate source file for new component
+    ComponentSpecification compSpec;
+    compSpec.typeName = typeName;
```

Comment:
not sure if needed, but subtype can no be specified here, so only "main" components can be added. Later typename should be capable of containing full type with some separator. Ex: "HydraulicOrific:MySpecialOrifice", but for now subtype is needed. But maybe not int this case because you can not add such component via dialog anyway I assume.